packages/ai-jsx/src/react/completion.tsx

```diff
@@ -77,40 +85,124 @@ export async function* UICompletion(
     const children =
       typeof serializedComponent.children === 'string' ? [serializedComponent.children] : serializedComponent.children;
 
-    return <Component>{children.map(toComponent)}</Component>;
+    if (serializedComponent.tag in validAIComponentsMap) {
+      if (serializedComponent.possiblyIncomplete) {
+        return `Loading ${serializedComponent.tag} ...`;
+      }
+      return (
+        <AI.jsx>{getOrCreateMemo(serializedComponent, <Component>{children.map(toComponent)}</Component>)}</AI.jsx>
+      );
+    }
+    return (
+      <AI.React>
+        <Component>{children.map(toComponent)}</Component>
+      </AI.React>
+    );
   }
 
   const Element: z.Schema = z.object({
-    name: z.string().refine((c) => Boolean(validComponents[c]), {
-      message: `Unknown component "name". Supported components: ${Object.keys(validComponents)}`,
+    tag: z.string().refine((c) => Boolean(validComponentsMap[c]), {
+      message: `Unknown component "tag". Supported components: ${Object.keys(validComponentsMap)}`,
     }),
-    children: z.array(z.union([z.string(), z.lazy(() => Element)])),
+    children: z.union([z.string(), z.array(z.union([z.string(), z.lazy(() => Element)]))]),
   });
+  const Elements = z.union([Element, z.array(Element)]);
```

Comment:
Variable names should be lower-cased unless they refer to TS types or JSX components.